src/main/java/org/apache/sysds/runtime/matrix/data/LibMatrixReorg.java

```diff
@@ -193,7 +193,7 @@ public static MatrixBlock transpose(MatrixBlock in, MatrixBlock out, int k, bool
 		if(in.isEmptyBlock(false) || ((long) in.rlen * (long) in.clen < PAR_NUMCELL_THRESHOLD) || k == 1 ||
 			(SHALLOW_COPY_REORG && !in.sparse && !out.sparse && (in.rlen == 1 || in.clen == 1)) ||
 			(in.sparse && !out.sparse && in.rlen == 1) || (!in.sparse && out.sparse && in.rlen == 1) ||
-			(!in.sparse && out.sparse)) {
+			(!in.sparse && out.sparse) || (in.nonZeros != -1 && in.nonZeros > 1000)) {
```

Comment:
Please double check this conditions - it looks like all large transpose operations are forced into single-threaded execution.